# CS 4661: Introduction to Data Science
## Jay Tandel


### Problem: Debt Prediction
In this question, we work with another dataset from the textbook of "An Introduction to Statistical Learning."

## A
Read the dataset file “Credit.csv” (from github using the following command), and assign it to a Pandas DataFrame:

df  = pd.read_csv("https://github.com/mpourhoma/CS4661/raw/master/Credit.csv")

In [1]:
# Importing the required packages and libraries
import pandas as pd
import numpy as np
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
from sklearn.model_selection import cross_val_score
from sklearn import preprocessing
from sklearn.linear_model import LinearRegression
from sklearn import metrics

In [2]:
# Reading the csv file from web, and store it in panda DataFrame
df = pd.read_csv("https://github.com/mpourhoma/CS4661/raw/master/Credit.csv")

## B
Check out the dataset. The “Credit” dataset includes the “balance” column (average credit card debt for a number of individuals) as target, as well as several features: age, cards (number of credit cards), education (years of education), income (in thousands of dollars), limit (credit limit), marital status, and rating (credit rating).

In [3]:
#print the imported dataset
print(df)

      Income  Limit  Rating  Cards  Age  Education  Married  Balance
0     14.891   3606     283      2   34         11        1      333
1    106.025   6645     483      3   82         15        1      903
2    104.593   7075     514      4   71         11        0      580
3    148.924   9504     681      3   36         11        0      964
4     55.882   4897     357      2   68         16        1      331
..       ...    ...     ...    ...  ...        ...      ...      ...
395   12.096   4100     307      3   32         13        1      560
396   13.364   3838     296      5   65         17        0      480
397   57.872   4171     321      5   67         12        1      138
398   37.728   2525     192      1   44         13        1        0
399   18.701   5524     415      5   64          7        0      966

[400 rows x 8 columns]


## C
Generate the feature matrix and target vector (target is “balance” in this dataset). Then, normalize (scale) the features (note: don’t normalize the target vector!). To normalize the data, you can simply use scale(X)  from sklearn.

In [4]:
# Creating the Feature Matrix for the dataset:

# create a python list of feature names that would like to pick from the dataset:
feature_cols = ['Income','Limit','Rating','Cards','Age','Education','Married']

# use the above list to select the features from the original DataFrame
X = df[feature_cols] 

print(X)


      Income  Limit  Rating  Cards  Age  Education  Married
0     14.891   3606     283      2   34         11        1
1    106.025   6645     483      3   82         15        1
2    104.593   7075     514      4   71         11        0
3    148.924   9504     681      3   36         11        0
4     55.882   4897     357      2   68         16        1
..       ...    ...     ...    ...  ...        ...      ...
395   12.096   4100     307      3   32         13        1
396   13.364   3838     296      5   65         17        0
397   57.872   4171     321      5   67         12        1
398   37.728   2525     192      1   44         13        1
399   18.701   5524     415      5   64          7        0

[400 rows x 7 columns]


In [5]:
# Normalize the feature matrix
X = preprocessing.scale(X)
print(X)

X = pd.DataFrame(X, columns=feature_cols)
print(X.head())

[[-0.86158299 -0.48999879 -0.46553881 ... -1.2576741  -0.78492991
   0.79539491]
 [ 1.72743711  0.82826106  0.82870309 ...  1.5284506   0.49658831
   0.79539491]
 [ 1.68675551  1.01478681  1.02931059 ...  0.88996369 -0.78492991
  -1.25723711]
 ...
 [ 0.35946155 -0.24491264 -0.21963285 ...  0.65778663 -0.46455035
   0.79539491]
 [-0.21280808 -0.95891584 -1.05441888 ... -0.67723146 -0.1441708
   0.79539491]
 [-0.75334493  0.34199278  0.38866085 ...  0.48365384 -2.06644812
  -1.25723711]]
     Income     Limit    Rating     Cards       Age  Education   Married
0 -0.861583 -0.489999 -0.465539 -0.699130 -1.257674  -0.784930  0.795395
1  1.727437  0.828261  0.828703  0.031032  1.528451   0.496588  0.795395
2  1.686756  1.014787  1.029311  0.761194  0.889964  -0.784930 -1.257237
3  2.946152  2.068440  2.110003  0.031032 -1.141586  -0.784930 -1.257237
4  0.302928  0.070012  0.013331 -0.699130  0.715831   0.816968  0.795395


In [6]:
# select label (the last column) from the DataFrame
y = df['Balance']

print(y)

0      333
1      903
2      580
3      964
4      331
      ... 
395    560
396    480
397    138
398      0
399    966
Name: Balance, Length: 400, dtype: int64


## D
Split the dataset into testing and training sets with the following parameters: test_size=0.24, random_state=9.

In [7]:
# Splitting the dataset into testing and training:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.24, random_state=9)


In [8]:
# print the training set:
print(X_train)
print('\n')
print(y_train)

       Income     Limit    Rating     Cards       Age  Education   Married
252  0.787669  1.677604  1.721730  0.031032  0.483654  -1.746069 -1.257237
320 -0.822151 -1.551027 -1.481519  0.031032  1.296274  -0.144171  0.795395
351  0.465938  0.175421  0.123341 -1.429291  0.889964  -1.425689  0.795395
244 -0.643686 -1.279046 -1.345623  0.761194  0.599742   0.496588 -1.257237
155 -0.729595 -1.463403 -1.371508  0.761194 -1.257674  -1.425689  0.795395
..        ...       ...       ...       ...       ...        ...       ...
251 -0.475704 -0.665246 -0.569078  1.491355  0.599742  -0.464550  0.795395
56   0.337530  0.002776  0.110399  2.951679  1.354318   1.457727  0.795395
126 -0.533857  0.345897  0.505143  1.491355 -0.328966   0.496588  0.795395
348 -0.903003 -1.562305 -1.572116  0.031032  0.831919   0.176209  0.795395
382 -0.579397  0.329413  0.349834 -1.429291 -1.315718  -1.425689 -1.257237

[304 rows x 7 columns]


252    1355
320       5
351     302
244       0
155       0
       ... 
25

In [9]:
t

       Income     Limit    Rating     Cards       Age  Education   Married
57   0.882214  0.018826  0.078043  0.761194  0.541698   0.176209  0.795395
32   2.527319  1.345762  1.346400 -0.699130 -0.445054  -0.144171 -1.257237
17  -0.247808 -0.155120 -0.103151  0.031032  0.773875   0.496588  0.795395
167 -0.449312 -1.044804 -1.145016 -1.429291 -1.083541   0.176209  0.795395
2    1.686756  1.014787  1.029311  0.761194  0.889964  -0.784930 -1.257237
..        ...       ...       ...       ...       ...        ...       ...
25  -0.884339 -0.178978 -0.187277  1.491355 -1.780072   0.816968  0.795395
397  0.359462 -0.244913 -0.219633  1.491355  0.657787  -0.464550  0.795395
62  -0.412977 -1.329365 -1.261498 -0.699130  1.238229  -2.066448 -1.257237
305 -0.589738 -1.219618 -1.229142 -0.699130 -0.328966   0.176209  0.795395
379  0.229803 -0.813599 -0.912052 -1.429291  1.644539  -1.105309  0.795395

[96 rows x 7 columns]


57     133
32     526
17     368
167      0
2      580
      ... 
25     67

## E
Use Linear Regression to train a linear model on the training set. Check the coefficients of the linear regression model. Which feature is the most important? Which feature is the least important?

In [10]:
# In the following line, "my_linreg" is instantiated as an "object" of LinearRegression "class". 

my_linreg = LinearRegression()

# fitting the model to the training data:
my_linreg.fit(X_train, y_train)

LinearRegression()

In [11]:
# printing Theta0 using attribute "intercept_":
print(my_linreg.intercept_)

# printing [Theta1, Theta2, Theta3] using attribute "coef_":
print(my_linreg.coef_)

print("Coeeficients of the linear Regression: ", my_linreg.coef_)

520.0391034603759
[-264.69553256  360.5601294   252.73642988   24.91707959  -17.23897686
   11.65810466   -9.82859046]
Coeeficients of the linear Regression:  [-264.69553256  360.5601294   252.73642988   24.91707959  -17.23897686
   11.65810466   -9.82859046]


In [12]:
# Features: Income     Limit    Rating     Cards       Age  Education   Married

print("Most important feature: Limit")
print("Least important feature: Married")

Most important feature: Limit
Least important feature: Married


## F
Predict “balance” for the users in testing set. Then, compare the predicted balance with the actual balance by calculating and reporting the RMSE (as we saw in lab tutorial 4).

In [13]:
# make predictions on the testing set
y_prediction = my_linreg.predict(X_test)

print(y_prediction)


[ 316.89813114  677.5093907   489.03669272  -49.82989971  706.42913216
  864.96954115  897.65752581  108.20297732 -213.73474279  834.94198677
   11.05109139 -227.34447869  472.01450822    4.54554458 -228.17349827
  956.21537845  896.91570524  821.97824267   18.37144232  857.31517013
 1028.30326279  699.22168497 1176.9426756   666.28578465  643.94064891
  690.63628696  589.35273903  -29.37237394  575.15444917  406.78468032
  845.81453327  828.30862904   82.2376206   953.40592514 -184.74828227
  525.84335889 1045.19205022  535.04104472  107.26753118 -172.8968111
  497.84037664 1155.35533939  429.08674601  403.49872552  143.72087313
  764.78737012  431.24114308 1304.79656588  475.00132057   95.75404622
 -215.69000296    5.62900657  313.88797405  859.9328932   792.27147016
 1029.90210771 1500.40180297 1078.48617624  713.37981473 1072.97948262
  -64.12187718  488.60063387   92.3601712   440.414465    277.13284667
 1075.41508762  258.10181743  773.89657462  278.12447916  409.51029492
  335.5

In [14]:
print(y_test)

57     133
32     526
17     368
167      0
2      580
      ... 
25     671
397    138
62       0
305      0
379      0
Name: Balance, Length: 96, dtype: int64


In [15]:
# Calculating "Mean Square Error" (MSE):
mse = metrics.mean_squared_error(y_test, y_prediction)

# Using numpy sqrt function to take the square root and calculate "Root Mean Square Error" (RMSE)
rmse = np.sqrt(mse)

print(rmse)

143.24649739642277


In [16]:
print("RMSE is: ",rmse)

RMSE is:  143.24649739642277


## G
Now, use 10-fold Cross-Validation to evaluate the performance of a linear regression in predicting the balance. Thus, rather than splitting the dataset into testing and training, use Cross-Validation to evaluate the regression performance. What is the RMSE when you use cross validation?

In [17]:
# Applying 10-fold cross validation with "linear regression" classifier:

# In the following line, "my_linreg1" is instantiated as an "object" of LinearRegression "class". 
my_linreg1 = LinearRegression()

# function cross_val_score performs Cross Validation:
mse_list = cross_val_score(my_linreg1, X, y, cv=10, scoring='neg_mean_squared_error')

print(mse_list)

[-23646.90415343 -32003.04401232 -35462.64435619 -37327.60719635
 -14341.32205939 -33628.37104224 -31631.99317834 -12491.00334951
 -20749.61212175 -23204.94743459]


In [18]:
# Notice that "cross_val_score" by default provides "negative" values for "mse" to clarify that mse is error.
# in order to calculate root mean square error (rmse), we have to make them positive!
mse_list_positive = -mse_list

# using numpy sqrt function to calculate rmse:
rmse_list = np.sqrt(mse_list_positive)
print(rmse_list)


[153.7754992  178.89394627 188.31527914 193.20353826 119.75525901
 183.38039983 177.85385343 111.76315739 144.04725656 152.331702  ]


In [19]:
# calculate the average RMSE as final result of cross validation:
print(rmse_list.mean())

160.33198910744136


In [20]:
print("RMSE with 10-fold cross validation: ", rmse_list.mean())

RMSE with 10-fold cross validation:  160.33198910744136


## Answer Summary

### E
#### Most important feature: Limit
#### Least important feature: Married

### F
#### RMSE is:  143.24649739642277
#### RMSE with 10-fold cross validation:  160.33198910744136